span_implementation.go

```diff
@@ -77,21 +78,83 @@ func (s *spanImpl) Tag(key, value string) {
 }
 
 func (s *spanImpl) Finish() {
+	d := time.Since(s.Timestamp)
 	if atomic.CompareAndSwapInt32(&s.mustCollect, 1, 0) {
-		s.Duration = time.Since(s.Timestamp)
-		if s.flushOnFinish {
+		s.mtx.Lock()
+		s.Duration = d
+		s.mtx.Unlock()
+
+		shouldRecord := true
+		if s.finishedSpanHandler != nil {
+			shouldRecord = s.finishedSpanHandler(&s.SpanModel)
+		}
+
+		if shouldRecord && s.flushOnFinish {
```

Comment:
I assume flushOnFinish is related to sampling bit. if so, sg

Reply:
Not sure that is the case. I think it is more for when you want to add more stuff after a span is finished. See https://github.com/openzipkin/zipkin-go/blob/master/span_options.go#L79

Reply:
gotcha. yeah I think this will want sampled local support before people start coding to it. Otherwise, later they might get a firehose and not notice. here's the docs from the java side.

```java
/**
 * Triggered on each finished span except when spans that are {@link Span#isNoop() no-op}.
 *
 * <p>{@link TraceContext#sampled() Sampled spans} hit this stage before reporting to Zipkin.
 * This means changes to the mutable span will reflect in reported data.
 *
 * <p>When Zipkin's reporter is {@link zipkin2.reporter.Reporter#NOOP} or the context is
 * unsampled, this will still receive spans where {@link TraceContext#sampledLocal()} is true.
 *
 * @see #alwaysSampleLocal()
 */
```

https://github.com/openzipkin/brave/blob/5a17fc018613958db00cc7b6951826e95f1b9d6c/brave/src/main/java/brave/handler/FinishedSpanHandler.java#L30